Ecoal (esterownik.pl) static fuel boiler and pump controller

matkor opened the pull request:
## Description:
Ecoal (esterownik.pl) static fuel boiler and pump controller interface.
Allows reading temp sensors and operate pumps connected to controller.

## Example entry for `configuration.yaml` :
```yaml
ecoal_boiler:
    host: 192.168.X.Y
    username: admin
    password: admin
    switches:
        monitored_conditions:
        - central_heating_pump
        - central_heating_pump2
    sensors:
        monitored_conditions:
        - outdoor_temp
        - indoor_temp
        - indoor2_temp
        - domestic_hot_water_temp
        - feedwater_out_temp
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io/pull/7707)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable (ecoaliface==0.4.0).
  - [X] New dependencies are only imported inside functions that use them.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

matkor committed:
Starting work on ecoal boiler controller iface.

matkor committed:
Sending some values/states to controller.

matkor committed:
Basic status parsing, and simple settings.

matkor committed:
Platform configuration.

matkor committed:
Merge branch 'dev' into ecoal_boiler

matkor committed:
Temp sensors seems be working.

matkor committed:
Switch from separate h/m/s to datetime.

matkor committed:
Vocabulary updates.

matkor committed:
secondary_central_heating_pump -> central_heating_pump2

matkor committed:
Pumps as switches.

matkor committed:
Optional enabling pumps via config.

matkor committed:
requests==2.20.1 added to REQUIREMENTS.

matkor committed:
Optional enabling temp sensors from configuration yaml.

matkor committed:
autopep8, black, pylint.

matkor committed:
flake8.

matkor committed:
pydocstyle

matkor committed:
All style checkers again.

matkor committed:
requests==2.20.1 required by  homeassistant.components.sensor.ecoal_boiler.

matkor committed:
Verify / set switches in update().
Code cleanup.

matkor committed:
script/lint + travis issues.

matkor committed:
Cleanup, imperative mood.

matkor committed:
pylint, travis.

matkor committed:
Updated .coveragerc.

frenck commented:
Could not find a related PR on our documentation repository. Adding `docs-missing` label.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+DOMAIN = "ecoal_boiler"
+
+CONF_HOSTNAME = "host"
```
`CONF_HOST` is available in `const.py`.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+DOMAIN = "ecoal_boiler"
+
+CONF_HOSTNAME = "host"
+CONF_LOGIN = "login"
```
Rename this to `username`.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+CONF_HOSTNAME = "host"
+CONF_LOGIN = "login"
+DEFAULT_LOGIN = "admin"
+CONF_PASSWORD = "password"
```
See above.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+#        vol.Optional(CONF_PASSWORD,
+#                        default=DEFAULT_PASSWORD): cv.string,
+#    })
+# })
```
Configuration needs to be validated.

matkor replied:
I tried to validate by included but commented out code:
```
CONFIG_SCHEMA = vol.Schema({
   DOMAIN: vol.Schema({
       vol.Required(CONF_HOST): cv.string,
       vol.Optional(CONF_USERNAME,
                       default=DEFAULT_USERNAME): cv.string,
       vol.Optional(CONF_PASSWORD,
                       default=DEFAULT_PASSWORD): cv.string,
   })
})
```

But it fails with
`Invalid config for [ecoal_boiler]: [homeassistant] is an invalid option for [ecoal_boiler].`
:( for config:

```
ecoal_boiler:
  host: 192.168.9.2
  username: admin
  password: admin
```

I will really appreciate if someone enlightens me what I am doing wrong here.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+async def async_setup(hass, config):
+    """Set up global ECOAL_CONTR same for sensors and switches."""
+    global ECOAL_CONTR
```
Don't use `GLOBAL` -> https://developers.home-assistant.io/docs/en/creating_component_code_review.html#3-component-platform-communication

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    _LOGGER.debug("async_setup(): config: %r", config)
+    from .http_iface import ECoalControler
+
+    # hass.states.set('hello.world', 'Paulus')_LOGGE
```
Stale comment.

matkor replied:
Fixed.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+async def async_setup(hass, config):
+    """Set up global ECOAL_CONTR same for sensors and switches."""
+    global ECOAL_CONTR
+    _LOGGER.debug("async_setup(): config: %r", config)
```
Doesn't look useful for a user. Reword it or remove it.

matkor replied:
Removed.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    conf = config.get(DOMAIN)
+    _LOGGER.debug(
+        "async_setup(): conf: %r  conf.keys(): %r", conf, conf.keys()
+    )
```
See above.

matkor replied:
Removed.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    passwd = conf.get(CONF_PASSWORD)
+    _LOGGER.debug(
+        "async_setup(): host: %r login: %r passwd: %r", host, login, passwd
+    )
```
See above

matkor replied:
Removed.

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+        "async_setup(): host: %r login: %r passwd: %r", host, login, passwd
+    )
+    ECOAL_CONTR = ECoalControler(host, login, passwd)
+    # _LOGGER.debug("async_setup(): ECOAL_CONTR: %r", ECOAL_CONTR)
```
See above.

matkor replied:
Removed.

fabaff commented on homeassistant/components/sensor/ecoal_boiler.py:
```diff
@@ -0,0 +1,101 @@
+  host: 192.168.1.123
+  login: admin
+  password: admin
+
```
This should go into the documentation.

matkor replied:
Docs PR created: https://github.com/home-assistant/home-assistant.io/pull/7707

fabaff commented on homeassistant/components/ecoal_boiler/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+Allows read various readings available in controller
+and set very basic switches.
+"""
```
Please refer to other component about the layout of the file header.

MartinHjelmare replied:
Yes, but say it's a component for the component.

matkor replied:
Is it OK in current version?

fabaff commented on homeassistant/components/ecoal_boiler/http_iface.py:
```diff
@@ -0,0 +1,602 @@
+"""
```
This is device specific code. It should be move to a Python module which is published on PyPI and hen imported by the component.

matkor replied:
Done as https://pypi.org/project/ecoaliface/ .

fabaff commented on homeassistant/components/sensor/ecoal_boiler.py:
```diff
@@ -0,0 +1,101 @@
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ["requests==2.20.1"]
```
`requests` is a core dependency. Just import it.

fabaff commented on homeassistant/components/sensor/ecoal_boiler.py:
```diff
@@ -0,0 +1,101 @@
+        self._name = name
+        self._status_attr = status_attr
+        self._state = None
+        self.update()
```
Don't call `self.update()` in the [constructor](https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#5-entity).

matkor committed:
Using configuration consts from homeassistant.const

matkor committed:
typo.

matkor committed:
Replace global ECOAL_CONTR with hass.data[DATA_ECOAL_BOILER].
Remove requests from REQUIREMENTS.

matkor committed:
Killed .update()/reread_update() in Entities __init__()s.
Removed debug/comments.

matkor committed:
Removed debug/comments.

matkor committed:
script/lint fixes.

matkor committed:
script/gen_requirements_all.py run.

matkor committed:
Travis fixes.

matkor committed:
Configuration now validated.

matkor committed:
Split controller code to separate package.

matkor committed:
Replace in module docs with link to https://home-assistant.io .

MartinHjelmare commented on .coveragerc:
```diff
@@ -90,6 +90,9 @@ omit =
     homeassistant/components/eight_sleep.py
     homeassistant/components/*/eight_sleep.py
 
+    homeassistant/components/ecoal_boiler/*
```
Update the path so that it matches the component module.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,46 @@
+    host = conf.get(CONF_HOST)
+    username = conf.get(CONF_USERNAME)
+    passwd = conf.get(CONF_PASSWORD)
+    ecoal_contr = ECoalController(host, username, passwd)
```
Check if credentials are correct here, otherwise `return False`.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,46 @@
+DOMAIN = "ecoal_boiler"
+DATA_ECOAL_BOILER = 'data_' + DOMAIN
+
+CONF_USERNAME = "username"
```
Import this from const.py.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,46 @@
+    from ecoaliface.simple import ECoalController
+
+    conf = config.get(DOMAIN)
+    host = conf.get(CONF_HOST)
```
Use `dict[key]` for required config keys and keys with default config schema values.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/sensor/ecoal_boiler.py:
```diff
@@ -0,0 +1,83 @@
+for tempsensor_id in SENSOR_IDS:
+    ENABLED_SCHEMA[vol.Optional(tempsensor_id)] = cv.string
+CONF_ENABLE = "enable"
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Instead of specifying a config section for each platform, keep the all the config in the component section, and load the platforms via discovery from the component.

Look at the octoprint component for the correct schema structure.

The platform schema here can be removed.

matkor replied:
You want me to move to parse pump/sensor configuration to component config and parse it in setup(hass, config) and call load_platform() for parsed pumps and sensors configs?
It is just moving configuration to different part of yaml, or I am missing other points here?

MartinHjelmare replied:
Yes, basically.

Having all of the config in the component section is the best option for components with platforms.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/sensor/ecoal_boiler.py:
```diff
@@ -0,0 +1,83 @@
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Rename `add_devices` to `add_entities`.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+for pump_id in PUMP_IDNAMES:
+    ENABLED_SCHEMA[vol.Optional(pump_id)] = cv.string
+CONF_ENABLE = "enable"
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
See above.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+        self._ecoal_contr = ecoal_contr
+        self._name = name
+        self._state_attr = state_attr
+        # NOTE: Convetion set_<attr> and status.<attr>
```
Convention?

This comment isn't clear to me what it means.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
See above.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+    @property
+    def should_poll(self) -> bool:
+        """Return if polling needed."""
+        return True
```
This is the default.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+    add_devices(switches, True)
+
+
+class EcoalSwitch(ToggleEntity):
```
Inherit from SwitchDevice in the switch component.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+        """Fetch state of pump without using cache."""
+        # Get status without using cache
+        self._ecoal_contr.get_status()
+        self.update()
```
This will be called at the end of each service call.

MartinHjelmare replied:
Yes, here is some more info:
https://developers.home-assistant.io/docs/en/dev_101_services.html

Services is what allows home assistant to change the state of a device that is actuatable. When we turn on a light entity in the frontend, home assistant calls the turn_on service with the corresponding entity_id in the service data, which in turn calls the turn_on method on the entity.

Entities that use home assistant built in polling, ie that have `should_poll` return `True` (this is the default) will have the `update' method be called at the end of the service call before updating the state.

matkor replied:
OK. Implemented. Thanks for excellent explanation.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+        self._contr_set_fun(1)
+        # Reread state without using cache.
+        self.reread_update()
+        self.schedule_update_ha_state()
```
This can be removed. It will be called at the end of each service call.

matkor replied:
Removed.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,108 @@
+        self._contr_set_fun(0)
+        # Reread state without using cache.
+        self.reread_update()
+        self.schedule_update_ha_state()
```
See above.

matkor committed:
Correct component module path in .coveragerc.
More vals from  const.py.
Use dict[key] for required config keys.
Check if credentials are correct during component setup.
Renamed add_devices to add_entities.

matkor committed:
Sensor/switch depends on ecoal_boiler component.
EcoalSwitch inherits from SwitchDevice.
Killed same as default should_poll().
Remove not neede schedule_update_ha_state() calls from turn_on/off.

matkor committed:
lint fixes.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,104 @@
+        self._ecoal_contr = ecoal_contr
+        self._name = name
+        self._state_attr = state_attr
+        # NOTE: Ecoalcotroller holds convention that same postfix is used
```
Ecoalcontroller

MartinHjelmare replied:
We can remove the `NOTE:` annotation, and just use regular comment `#`.

matkor replied:
Done

matkor committed:
Move sensors/switches configuration to component setup.

matkor committed:
Lint fixes.

matkor committed:
Invalidating ecoal iface cache instead of force read in turn_on/off().

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -21,31 +22,77 @@
+CONF_SWITCHES = 'switches'
+SWITCHES_SCHEMA = {}
+for pump_id in PUMP_IDNAMES:
+    SWITCHES_SCHEMA[vol.Optional(pump_id)] = cv.string
```
We don't need to allow a custom name per sensor or switch type. Change this validation to a list of types and just check that the configured sensor types are in the list of allowed types.

Default names should be set by the platforms from the sensor or switch types.

matkor replied:
There is no good default.
Switches as pump can be configured (internally in controller) to perform different tasks depending on heating configuration. 
Sensors can be placed in different places(rooms) so there should be a way to set a name visible in GUI.

MartinHjelmare replied:
Names can always be customized. If a unique_id is implemented it can be done very easily.

As long as we can add a separate name for each entity, that will be good enough.

matkor replied:
OK, so assuming I would like to set name of given pump how that should be done in config?
Component config would be
```
ecoal_boiler:
    switches:
        central_heating_pump
```
where should go name of  _central_heating_pump_  (like _Wall radiators_ ) ?

matkor replied:
Entity names are now hardcoded.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,98 @@
+    if sensors:
+        load_platform(hass, 'sensor', DOMAIN, sensors, hass_config)
+
+    if ecoal_contr.version is None:
```
This should be done as early as possible.

matkor replied:
Done.

matkor committed:
Fail component setup before adding any platform entities.
Kill NOTE.

matkor committed:
Disallow setting entity names from config file, use code defined default names.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,97 @@
+
+
+CONF_SWITCHES = 'switches'
+SWITCHES_SCHEMA = []
```
Look at the rainmachine component for the correct schema.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,97 @@
+    SWITCHES_SCHEMA.append(vol.Optional(pump_id))
+
+CONF_SENSORS = 'sensors'
+SENSORS_SCHEMA = []
```
See above.

matkor replied:
1. What "type" do you wish to have for temperature sensors? 
```
SENSORS = {
    TYPE_WHAT_HERE: ('Freeze Protect Temperature', 'mdi:thermometer', '°C'),
}
```
2. How should I convert
`SWITCH_SCHEMA = vol.Schema({vol.Optional(CONF_ZONE_RUN_TIME): cv.positive_int})`
to list switches offered by controller?

Or could you please provide correct schema, I will adapt current code to parse it?

MartinHjelmare replied:
Type should be a lowercase snakecase name that matches what the sensor does.

Here's the sensor schema eg:
```py
SENSOR_SCHEMA = vol.Schema({
    vol.Optional(CONF_MONITORED_CONDITIONS, default=list(AVAILABLE_SENSORS)):
        vol.All(cv.ensure_list, [vol.In(AVAILABLE_SENSORS)])
})
```
Here's the switch schema:
```py
SWITCH_SCHEMA = vol.Schema({
    vol.Optional(CONF_MONITORED_CONDITIONS, default=list(AVAILABLE_PUMPS)):
        vol.All(cv.ensure_list, [vol.In(AVAILABLE_PUMPS)])
})
```
Adjust as necessary. Eg should we have a default at all or should the monitored conditions key be required?

MartinHjelmare replied:
`AVAILABLE_SENSORS` dict looks good, so that doesn't need changing, unless you want to add more items than name.

matkor replied:
Please check recent version, as I am little confused with that style of config.  Config key `CONF_MONITORED_CONDITIONS` looks like not needed to me, but I assume it's common in other configs.
Default all for both pumps and sensors is OK.

matkor committed:
Rework configuration file to use monitored_conditions like in rainmachine component.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,96 @@
+    passwd = conf[CONF_PASSWORD]
+    ecoal_contr = ECoalController(host, username, passwd)
+    if ecoal_contr.version is None:
+        # Wrong credentials nor network config
```
Log an error here too.

matkor replied:
Done.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,83 @@
+    """Set up switches based on ecoal interface."""
+    ecoal_contr = hass.data[DATA_ECOAL_BOILER]
+    switches = []
+    for pump_id in discovery_info:     # pylint: disable=W0621
```
We can remove the pylint disable.

matkor replied:
Killed.

MartinHjelmare reviewed (commented):
Nice!

matkor committed:
Killed pylint exception.
Log error when connection to controller fails.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,98 @@
+    if ecoal_contr.version is None:
+        # Wrong credentials nor network config
+        _LOGGER.error("Unable to read controller status from %s@%s"
+                      " (wrong host/credentials).", username, host, )
```
We don't end logging messages with period.

MartinHjelmare commented on homeassistant/components/ecoal_boiler.py:
```diff
@@ -0,0 +1,98 @@
+        return False
+    _LOGGER.debug("Detected controller version: %r @%s",
+                  ecoal_contr.version, host, )
+    # Creating ECoalController instance makes HTTP request to controller.
```
Move this comment up above where we create the controller.

MartinHjelmare commented on homeassistant/components/sensor/ecoal_boiler.py:
```diff
@@ -0,0 +1,61 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the ecoal sensors."""
+    devices = []
```
Add a guard clause here that checks if discovery_info is None and return if so.

MartinHjelmare commented on homeassistant/components/switch/ecoal_boiler.py:
```diff
@@ -0,0 +1,83 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up switches based on ecoal interface."""
+    ecoal_contr = hass.data[DATA_ECOAL_BOILER]
```
See above.

MartinHjelmare commented:
Please update the config example in the PR description, to match the new config schema.

matkor committed:
A few fixes.

matkor commented:
> Please update the config example in the PR description, to match the new config schema.

Done.
I assume we have config schema settled, so I can update documentation PR as well?

matkor committed:
Linted.

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Can be merged when build passes.
